sdk/metric/batcher/defaultkeys/defaultkeys.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package defaultkeys // import "go.opentelemetry.io/otel/sdk/metric/batcher/defaultkeys"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+)
+
+type (
+	Batcher struct {
+		selector export.AggregationSelector
+		lencoder export.LabelEncoder
+		stateful bool
+		dki      dkiMap
+		agg      aggMap
+	}
+
+	aggEntry struct {
+		descriptor *export.Descriptor
+		labels     export.Labels
+		aggregator export.Aggregator
+	}
+
+	dkiMap map[*export.Descriptor]map[core.Key]int
+	aggMap map[string]aggEntry
+
+	producer struct {
+		aggMap   aggMap
+		lencoder export.LabelEncoder
+	}
+)
+
+var _ export.Batcher = &Batcher{}
+var _ export.Producer = &producer{}
+
+func New(selector export.AggregationSelector, lencoder export.LabelEncoder, stateful bool) *Batcher {
+	return &Batcher{
+		selector: selector,
+		lencoder: lencoder,
+		dki:      dkiMap{},
+		agg:      aggMap{},
+		stateful: stateful,
+	}
+}
+
+func (b *Batcher) AggregatorFor(descriptor *export.Descriptor) export.Aggregator {
+	return b.selector.AggregatorFor(descriptor)
+}
+
+func (b *Batcher) Process(_ context.Context, desc *export.Descriptor, labels export.Labels, agg export.Aggregator) error {
+	keys := desc.Keys()
+
+	// Cache the mapping from Descriptor->Key->Index
+	ki, ok := b.dki[desc]
+	if !ok {
+		ki = map[core.Key]int{}
+		b.dki[desc] = ki
+
+		for i, k := range keys {
+			ki[k] = i
+		}
+	}
+
+	// Compute the value list.  Note: Unspecified values become
+	// empty strings.  TODO: pin this down, we have no appropriate
```

Comment:
I decided to leave this as-is for the moment. It's true that INVALID can be used in this way, but it will complicate this PR to do it correctly. The issue is that all label encoders should then distinguish this unspecified case, which means adding an extra byte to the encoding somehow. Also this is currently being debated at the spec and protocol level, so it should wait.

Reply:
I filed issue #320